### PR TITLE
support for CollateExpression

### PR DESCRIPTION
**Pull request by kanha-gupta**
Ref #24200 
Changes made -
1. Optimized MySQL grammar file for Collate expression
2. Created a Converter file for collate expression
3. Since Calcite only support of ISO type as Collate naming format, The approach for this PR is avoiding the use of Calcite's SqlCollation class & Using an alternative way of calling SqlBasicCall class to perform the task.

Before committing this PR, I'm sure that I have checked the following options:
- [ ] My code follows the [code of conduct](https://shardingsphere.apache.org/community/en/involved/conduct/code/) of this project.
- [ ] I have self-reviewed the commit code.
- [ ] I have (or in comment I request) added corresponding labels for the pull request.
- [ ] I have passed maven check locally : `./mvnw clean install -B -T1C -Dmaven.javadoc.skip -Dmaven.jacoco.skip -e`.
- [ ] I have made corresponding changes to the documentation.
- [ ] I have added corresponding unit tests for my changes.

### CODE REVIEW

**Comment by kanha-gupta** [comment]
@strongduanmu Sir please review, Collate naming logic requires work :)

**Comment by strongduanmu** [comment]
> @strongduanmu Sir please review, Collate naming logic requires work :)

Good job, I will review this pr.

File: parser/sql/statement/src/main/java/org/apache/shardingsphere/sql/parser/sql/common/segment/dml/expr/CollateExpression.java [hunk 1]
```diff
@@ -33,6 +36,9 @@ public final class CollateExpression implements ExpressionSegment {
     private final int stopIndex;
     
     private final SimpleExpressionSegment collateName;
+
+    @Nullable
```
Comment (strongduanmu) [inline parser/sql/statement/src/main/java/org/apache/shardingsphere/sql/parser/sql/common/segment/dml/expr/CollateExpression.java:40]: Please remove @Nullable annotation, you can use following code.

```java
public Optional< ColumnSegment> getTableName() {
...
}
```

File: parser/sql/statement/src/main/java/org/apache/shardingsphere/sql/parser/sql/common/segment/dml/expr/CollateExpression.java [hunk 2]
```diff
@@ -33,6 +36,9 @@ public final class CollateExpression implements ExpressionSegment {
...
+    private final ColumnSegment tableName;
```
Comment (strongduanmu) [inline parser/sql/statement/src/main/java/org/apache/shardingsphere/sql/parser/sql/common/segment/dml/expr/CollateExpression.java:41]: Besides, why use tableName for ColumnSegment?

File: kernel/sql-federation/core/src/main/java/org/apache/shardingsphere/sqlfederation/compiler/converter/segment/expression/impl/CollateExpressionConverter.java [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+import org.apache.calcite.sql.SqlIdentifier;
+import org.apache.calcite.sql.SqlNode;
+
+import org.apache.calcite.sql.parser.SqlParserPos;
+import org.apache.shardingsphere.sql.parser.sql.common.segment.dml.expr.CollateExpression;
+import org.apache.shardingsphere.sqlfederation.compiler.converter.segment.SQLSegmentConverter;
+
+import java.util.Optional;
+
+public final class CollateExpressionConverter implements SQLSegmentConverter<CollateExpression, SqlNode> {
```
Comment (strongduanmu) [inline kernel/sql-federation/core/src/main/java/org/apache/shardingsphere/sqlfederation/compiler/converter/segment/expression/impl/CollateExpressionConverter.java:31]: Please add java doc for CollateExpressionConverter.

**Review by strongduanmu — CHANGES_REQUESTED** [review]

**Comment by kanha-gupta** [comment]
@strongduanmu Please review, Collate queries are being parsed now :)

**Review by strongduanmu — APPROVED** [review]
I like this solution, merged.

**Pull request merged**